File: 002.Artificial_Intelligence/012.AIAgent/006.第 6 课 Spring AI/workspace/day_04_spring_ai_rag_tool/src/main/java/com/demo/springai/opsagent/infrastructure/tool/OpsTools.java
package com.demo.springai.opsagent.infrastructure.tool;

import com.demo.springai.opsagent.infrastructure.ssh.LinuxSshExecutor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.ai.tool.annotation.Tool;
import org.springframework.ai.tool.annotation.ToolParam;
import org.springframework.stereotype.Component;

@Component
public class OpsTools {

    private static final Logger log = LoggerFactory.getLogger(OpsTools.class);

    private final LinuxSshExecutor sshExecutor;

    public OpsTools(LinuxSshExecutor sshExecutor) {
        this.sshExecutor = sshExecutor;
    }

    @Tool(description = """
            查询已配置 Linux 主机的磁盘使用情况（执行 df -h）。
            当前目标主机在配置中指定，hostId 填 linux-231 即可。
            """)
    public String queryDiskUsage(
            @ToolParam(description = "主机标识，填 linux-231") String hostId) {
        log.info("[Tool 调用] queryDiskUsage, hostId={}", hostId);
        return sshExecutor.runReadOnlyCommand("df -h");
    }

    @Tool(description = """
            查询已配置 Linux 主机内存占用最高的进程 Top3（只读 ps 命令）。
            """)
    public String queryTopMemoryProcesses(
            @ToolParam(description = "主机标识，填 linux-231") String hostId) {
        log.info("[Tool 调用] queryTopMemoryProcesses, hostId={}", hostId);
        return sshExecutor.runReadOnlyCommand("ps aux --sort=-%mem | head -n 4");
    }

    @Tool(description = "查询已配置 Linux 主机内存与 swap 使用情况（free -h）")
    public String queryMemoryUsage(
            @ToolParam(description = "主机标识，填 linux-231") String hostId) {
        log.info("[Tool 调用] queryMemoryUsage, hostId={}", hostId);
        return sshExecutor.runReadOnlyCommand("free -h");
    }

    @Tool(description = "查看 /var/log 下各子目录占用空间，辅助定位磁盘满是否由日志引起")
    public String queryVarLogDiskUsage(
            @ToolParam(description = "主机标识，填 linux-231") String hostId) {
        log.info("[Tool 调用] queryVarLogDiskUsage, hostId={}", hostId);
        return sshExecutor.runReadOnlyCommand("du -sh /var/log/* 2>/dev/null | sort -hr | head -n 5");
    }
}

File: 002.Artificial_Intelligence/012.AIAgent/002.tool/workspace/day_04_connection_manager/app/infrastructure/tool/ops_tools.py
# -*- coding: utf-8 -*-
"""运维 Tool：host_id = 连接管理中的 label。"""
from __future__ import annotations

from langchain_core.tools import tool

from app.infrastructure.ssh import LinuxSshExecutor


def build_ops_tools(ssh: LinuxSshExecutor) -> list:
    @tool
    def query_disk_usage(host_id: str) -> str:
        """查询 Linux 主机磁盘使用情况（df -h）。host_id 填连接管理中的 label，如 linux-231。"""
        print(f"[Tool] query_disk_usage host_id={host_id}")
        return ssh.run_read_only_command("df -h", host_id)

    @tool
    def query_top_memory_processes(host_id: str) -> str:
        """查询 Linux 主机内存占用最高进程 Top3。host_id 填连接管理中的 label。"""
        print(f"[Tool] query_top_memory_processes host_id={host_id}")
        return ssh.run_read_only_command("ps aux --sort=-%mem | head -n 4", host_id)

    @tool
    def query_memory_usage(host_id: str) -> str:
        """查询 Linux 主机内存与 swap（free -h）。host_id 填连接管理中的 label。"""
        print(f"[Tool] query_memory_usage host_id={host_id}")
        return ssh.run_read_only_command("free -h", host_id)

    @tool
    def query_var_log_disk_usage(host_id: str) -> str:
        """查看 /var/log 子目录占用。host_id 填连接管理中的 label。"""
        print(f"[Tool] query_var_log_disk_usage host_id={host_id}")
        return ssh.run_read_only_command(
            "du -sh /var/log/* 2>/dev/null | sort -hr | head -n 5", host_id
        )

    @tool
    def query_logged_in_users(host_id: str) -> str:
        """查询 Linux 主机当前有哪些用户已登录（who）。host_id 填连接管理中的 label。"""
        print(f"[Tool] query_logged_in_users host_id={host_id}")
        return ssh.run_read_only_command("who", host_id)

    return [
        query_disk_usage,
        query_top_memory_processes,
        query_memory_usage,
        query_var_log_disk_usage,
        query_logged_in_users,
    ]
